File: java/src/main/java/co/rerout/sdk/model/BatchLinkResult.java
/*
 * rerout-java — Official Java SDK for the Rerout branded-link API.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk.model;

/**
 * The outcome of a single link in a batch create. {@code index} ties the result
 * back to its position in the submitted {@code links} array. On success
 * {@code code} is set; on failure {@code error} carries the reason. Instances
 * are immutable.
 */
public final class BatchLinkResult {

    private final int index;

    private final boolean ok;

    private final String code;

    private final String error;

    /**
     * Creates a {@code BatchLinkResult}.
     *
     * @param index the position of this link in the submitted batch
     * @param ok    whether the link was created
     * @param code  the created link code, or {@code null} on failure
     * @param error the failure reason, or {@code null} on success
     */
    public BatchLinkResult(int index, boolean ok, String code, String error) {
        this.index = index;
        this.ok = ok;
        this.code = code;
        this.error = error;
    }

    /** {@return the position of this link in the submitted batch} */
    public int getIndex() {
        return index;
    }

    /** {@return whether the link was created} */
    public boolean isOk() {
        return ok;
    }

    /** {@return the created link code, or {@code null} on failure} */
    public String getCode() {
        return code;
    }

    /** {@return the failure reason, or {@code null} on success} */
    public String getError() {
        return error;
    }
}

File: php/src/Models/BatchLinkResult.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

use InvalidArgumentException;

/**
 * The outcome of a single link in a batch create. `index` ties the result back
 * to its position in the submitted `links` array. On success `code` is set; on
 * failure `error` carries the reason.
 */
final readonly class BatchLinkResult
{
    /**
     * @param int         $index Position of this link in the submitted batch.
     * @param bool        $ok    Whether the link was created.
     * @param string|null $code  Created link code, when `ok`.
     * @param string|null $error Failure reason, when not `ok`.
     */
    public function __construct(
        public int $index,
        public bool $ok,
        public ?string $code = null,
        public ?string $error = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $index = $data['index'] ?? null;
        if (!is_int($index)) {
            throw new InvalidArgumentException("Expected int field 'index' in BatchLinkResult payload.");
        }

        $ok = $data['ok'] ?? null;
        if (!is_bool($ok)) {
            throw new InvalidArgumentException("Expected bool field 'ok' in BatchLinkResult payload.");
        }

        $code = $data['code'] ?? null;
        if ($code !== null && !is_string($code)) {
            throw new InvalidArgumentException("Expected nullable string field 'code' in BatchLinkResult payload.");
        }

        $error = $data['error'] ?? null;
        if ($error !== null && !is_string($error)) {
            throw new InvalidArgumentException("Expected nullable string field 'error' in BatchLinkResult payload.");
        }

        return new self(index: $index, ok: $ok, code: $code, error: $error);
    }
}
